l1-contracts/contracts/governance/PermanentRestriction.sol

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.24;
+
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+
+import {Call} from "./Common.sol";
+import {IRestriction} from "./IRestriction.sol";
+import {IChainAdmin} from "./IChainAdmin.sol";
+import {IBridgehub} from "../bridgehub/IBridgehub.sol";
+import {IZkSyncHyperchain} from "../state-transition/chain-interfaces/IZkSyncHyperchain.sol";
+import {IAdmin} from "../state-transition/chain-interfaces/IAdmin.sol";
+
+import {IPermanentRestriction} from "./IPermanentRestriction.sol";
+
+/// @title PermanentRestriction contract
+/// @notice This contract should be used by chains that wish to guarantee that certain security
+/// properties are preserved forever.
+/// @dev To be deployed as a transparent upgradable proxy, owned by a trusted decentralized governance.
+/// @dev Once of the instances of such contract is to ensure that a ZkSyncHyperchain is a rollup forever.
+contract PermanentRestriction is IRestriction, IPermanentRestriction, Ownable2Step {
+    /// @notice The address of the Bridgehub contract.
+    IBridgehub public immutable BRIDGE_HUB;
+
+    /// @notice The mapping of the allowed admin implementations.
+    mapping(bytes32 implementationCodeHash => bool isAllowed) public allowedAdminImplementations;
+
+    /// @notice The mapping of the allowed calls.
+    mapping(bytes4 selector => mapping(bytes allowedCalldata => bool isAllowed)) public allowedCalls;
```

Comment:
You don't need both `selector` and `allowedCalldata`, do you? `allowedCalldata` contains `selector`

Reply:
`setAllowedData` is also error prone due to this invariant. if the first 4 bytes in `data` is not the same as `selector` then allowed call would never be executed